Issue449 optimize physics tend config

ahijevyc opened the pull request:
This pull request optimizes the physics tendency code in the fv3_physics_tend module.

1. Documentation Updates:

+ Updated several figures in the [figure](https://github.com/dtcenter/METplotpy/tree/issue449_optimize_physics_tend_config/docs/Users_Guide/figure) directory with blue-white-red colormap.
+ Removed outdated sample data link from [fv3_physics.rst](https://github.com/dtcenter/METplotpy/tree/issue449_optimize_physics_tend_config/docs/Users_Guide/fv3_physics.rst).
+ Simplified usage instructions for planview_fv3.py, vert_profile_fv3.py, and cross_section_vert.py.

2. Code Refactoring:

+ Refactored cross_section_vert.py, planview_fv3.py, and vert_profile_fv3.py to streamline argument parsing and configuration handling.
+ Moved common functions to physics_tend.py for better code reuse.

3. Configuration Enhancements:

+ New default settings for tendency variables, geographic extent, color maps, time window, valid time, and more.
+ Updated fv3_physics_tend_defaults.yaml for FV3 output `/glade/campaign/mmm/parc/ahijevyc/METplotpy/DavidA_fv3_history_etc_files`. These data have tendencies that balance the actual change in state variables (i.e. the residual is close to zero).
+ Added a configuration file UFS.yaml to handle FV3 UFS output in May Wong's campaign storage. These data have tendencies that are never reset to zero during the model run nor averaged over the time interval between output. They also have separate DataArrays for each vertical level. For these data, the configuration switch `tendencies_were_zeroed_and_averaged_after_every_output` must be set to `True` and the function `physics_tend.get_fv3ds` is used to stack DataArrays for the same state variable (t, u, v, or q) but different pressure levels in the vertical.

4. Performance Improvements:

+ Optimized the handling of large datasets by using chunking and lazy loading where appropriate.

5. TODO:

 - Write test for FV3 UFS output in May Wong's campaign storage.
 - Pass SonarQube Quality Gate check.

6. Pull Request Testing
<input disabled="" type="checkbox"> Describe testing already performed for these changes:</br>

+ new [Jupyter notebook](https://github.com/dtcenter/METplotpy/blob/issue449_optimize_physics_tend_config/metplotpy/contributed/fv3_physics_tend/fv3_physics_tend.ipynb) to plot plan views, vertical cross section, and vertical profiles of tendency using multiple input formats.
+ Plan view, vertical profile, and cross-sections all look the same, aside from the new colormap.
+ Plan view, vertical profile, and cross-section work for `/glade/campaign/mmm/parc/mwong/ufs-mrw/2020040812.F024.C768/fv3_history2d.tile3.nc` 
- Uploaded new test data from `/glade/campaign/mmm/parc/ahijevyc/METplotpy/DavidA_fv3_history_etc_files`.

ahijevyc committed:
get args from config yaml

Moved planview, vert_profile, and cross_section_vert out of main().

Moved optional command line args to config file.

try to show requisite yaml mods

space after .rst code-block

add tab to .rst code-block

removed extraneous space

adapted cross_section_vert

ahijevyc committed:
fix example cmd

with shapefile in config file not argument.
same with xmin and xmax

ahijevyc committed:
moved lots of optional args here

ahijevyc committed:
clean up args

No more argparse args to functions.

Save image to TMPDIR

override config file with kwargs.

ahijevyc committed:
better colormap

ahijevyc committed:
remove bad link and description of sample data

ahijevyc committed:
common code moved to physics_tend.get_datetimeindex

ahijevyc committed:
Handle UFS runs

in /glade/campaign/mmm/parc/mwong/ufs-mrw

They are different from previous history file.

vertical pressure levels are split into multiple variables, so they must be combined. Tendencies are accumulated from initialization without resetting to zero along the way. They aren't scaled by how much time has gone by since the most recent reset.

ahijevyc committed:
assert history and grid match

use regex to match FV3-style filename

coordinate reference system option in config file for cross_section_vert

ahijevyc committed:
RdBu_r for all test figures

ahijevyc committed:
str(history_file)

zoom in on vertical_profile x-axis in test case figure

ahijevyc committed:
put statevarname and tendencytype in config file

ahijevyc committed:
added statevarname and tendencytype

ahijevyc committed:
handle subtract

in physics_tend.get_fv3ds. Move to else-block in calling function.

ahijevyc committed:
physics_tend.prepare_ds()

will prepage a single dataset with history and grid spec, given the paths to both. Feed it to planview, vert_profile, and cross_section_vert.

Remove `subtract` option. It is ugly to do it in the functions, and easy to do beforehand.

ahijevyc committed:
less logging

moved shp inset to left side to avoid covering up power-of-ten in x-axis tick label

ahijevyc committed:
refactor get_fv3ds()

rename `fv3` configuration dictionary -> `config`

rename `time_after_time0` -> `time_after_twindow_start`

document get_fv3ds()
        This function handles the fv3 UFS forecasts made by May Wong
        in campaign storage
        It subtracts tendency values across a time window.
        Leaves state variables alone
        Stacks in the vertical
        Now it matches the older fv3 history files METplotpy was
        designed for.

make sure "validtime" and "twindow" are consistent between configuration
dictionary and Dataset .

ahijevyc committed:
Merge branch 'develop' into issue449_optimize_physics_tend_config

ahijevyc committed:
fixed nested quotes

ahijevyc committed:
Don't use old history file anymore

ahijevyc committed:
reproduce figures from older documentation in User Guide

ahijevyc commented:
Hi @bikegeek . I refactored the contributed code for fv3 tendency plotting. In addition to moving command-line options to the configuration file, I rewrote a lot of the code. It still works on the older test data and a new input format. I'd like to change the sample data tar file to contain fv3 output for which the physics tendencies properly balance the actual change in temperature, wind, and moisture.
The usual tests started once I submitted this PR, but the SonarQube test is failing, and I don't think it is getting to the tests for fv3 tendency. I'll look through the contributor's guide to look for hints.

bikegeek commented:
You will also need to update the data for the use case, following the
instructions in the Contributor's Guide:
https://metplus.readthedocs.io/en/latest/Contributors_Guide/add_use_case.html

Refer to section 9.5.1



On Mon, Mar 3, 2025 at 11:42 AM David Ahijevych ***@***.***>
wrote:

> Hi @bikegeek <https://github.com/bikegeek> . I refactored the contributed
> code for fv3 tendency plotting. In addition to moving command-line options
> to the configuration file, I rewrote a lot of the code. It still works on
> the older test data and a new input format. I'd like to change the sample
> data tar file to contain fv3 output for which the physics tendencies
> properly balance the actual change in temperature, wind, and moisture.
> The usual tests started once I submitted this PR, but the SonarQube test
> is failing, and I don't think it is getting to the tests for fv3 tendency.
> I'll look through the contributor's guide to look for hints.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplotpy/pull/492#issuecomment-2695250853>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHVYIF2BQ5BSZHFYMZ32SSPABAVCNFSM6AAAAABYHUC65SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJVGI2TAOBVGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: ahijevyc]*ahijevyc* left a comment (dtcenter/METplotpy#492)
> <https://github.com/dtcenter/METplotpy/pull/492#issuecomment-2695250853>
>
> Hi @bikegeek <https://github.com/bikegeek> . I refactored the contributed
> code for fv3 tendency plotting. In addition to moving command-line options
> to the configuration file, I rewrote a lot of the code. It still works on
> the older test data and a new input format. I'd like to change the sample
> data tar file to contain fv3 output for which the physics tendencies
> properly balance the actual change in temperature, wind, and moisture.
> The usual tests started once I submitted this PR, but the SonarQube test
> is failing, and I don't think it is getting to the tests for fv3 tendency.
> I'll look through the contributor's guide to look for hints.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplotpy/pull/492#issuecomment-2695250853>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4UJHVYIF2BQ5BSZHFYMZ32SSPABAVCNFSM6AAAAABYHUC65SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJVGI2TAOBVGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ahijevyc committed:
works with oldest test data

ahijevyc commented:
I attached new use case data here based on [discussion 954](https://github.com/dtcenter/METplus/discussions/954)

[fv3_physics_tendency.tgz](https://github.com/user-attachments/files/19131646/fv3_physics_tendency.tgz)

To be used with configuration file [use_case.yaml](https://github.com/dtcenter/METplotpy/blob/issue449_optimize_physics_tend_config/metplotpy/contributed/fv3_physics_tend/use_case.yaml)

ahijevyc committed:
handles 'balanced' tendencies

ahijevyc committed:
use smaller test files in notebook

ahijevyc committed:
Merge branch 'develop' into issue449_optimize_physics_tend_config

ahijevyc committed:
use config dict instead of saving new var

ahijevyc committed:
Update Users Guide

fewer cmd line args
update figures with lower-res test data

Only one usage: section (cause all 3 programs share parse_args function
now).

--debug cmd line option
force logging level change

ahijevyc committed:
tweak examples

ahijevyc committed:
Replace old default config

revise user guide on tendency varnames

bikegeek commented:
converting to draft mode until further commits have stopped.

ahijevyc committed:
WORKING_DIR->DATA_DIR

ahijevyc committed:
fix usage wording

ahijevyc commented:
I have some test things that I can do before turning it over. I’ll let you know when I’m done committing.DaveOn Mar 7, 2025, at 10:30 AM, MWin ***@***.***> wrote:﻿
  converting to draft mode until further commits have stopped.—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>



bikegeek left a comment (dtcenter/METplotpy#492)
converting to draft mode until further commits have stopped.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you authored the thread.Message ID: ***@***.***>

ahijevyc committed:
update tests

ahijevyc committed:
update README

ahijevyc committed:
remove 3 vert_profile tests

difference plot
novel filename
novel output directory

ahijevyc committed:
duplicated code to physics_tend.assert_times

ahijevyc committed:
unskip planvew tests

ahijevyc committed:
use config lat_name/lon_name

ahijevyc committed:
Moved repeated code to physics_tend

removed "resid" config switch
restored latt and lont to places that I tried to use config["lat_name"]
and config["lon_name"].

ahijevyc commented:
Finished committing. The use case data are in this tar file:
[fv3_physics_tendency.tgz](https://github.com/user-attachments/files/19145136/fv3_physics_tendency.tgz) Still needs to be uploaded by someone with permission to a location that unit tests can read.

ahijevyc commented:
PR still lacks test coverage of new code, so switched PR to draft. I need help with unit testing.

bikegeek committed:
Merge branch 'develop' of https://github.com/dtcenter/METplotpy into issue449_optimize_physics_tend_config

bikegeek committed:
Address FutureWarning in to_datetimeindex.  Currently, the default time unit is "ns" but will be replaced to "us". Explicitly retain the "ns" time units via the time_unit keyword

bikegeek committed:
Common fixtures that can be used across different tests.

bikegeek committed:
Config file with settings for running planview plot tests

bikegeek committed:
Modified test to utilize updated interfaces, no longer requiring test runners to create the command line commands to generate the planview plot.

bikegeek committed:
No longer required to run the tests

bikegeek committed:
Skip tests so the Github Actions CI don't run them, the data files are too large for Github

bikegeek committed:
Update test_planview.py

remove the redundant import statement for xarray

bikegeek committed:
Updates to enable reusing the same config file and setup

bikegeek committed:
Updated comments for description to indicate that this is used by all three plot types

bikegeek committed:
Moved functionality from test_planview into conftest for re-use by other tests

bikegeek committed:
Universal config file indicating where the common data files for all tests reside

bikegeek committed:
updated to use functions in conftest.py

bikegeek committed:
Clean up of unused methods and moved cleanup() from the test_planview.py to enable use by other tests

bikegeek committed:
Clean up, moved cleanup() to the local conftest.py

bikegeek committed:
Refactored to accommodate the updates to the underlying code. No longer need to test via command line commands.

bikegeek committed:
Updated imports to use cleanup() from conftest.py

bikegeek committed:
no longer needed

bikegeek committed:
No longer needed

bikegeek committed:
No longer needed.

bikegeek committed:
Skip these tests in GHA, the data files are too large.

bikegeek committed:
Merge branch 'issue449_optimize_physics_tend_config' of https://github.com/dtcenter/METplotpy into issue449_optimize_physics_tend_config

bikegeek committed:
Fix conflicts

bikegeek committed:
No longer needed

bikegeek committed:
Add setting for the plotting source path

bikegeek committed:
Updates to tests to use re-usable code in conftest.py, cleaning up

bikegeek committed:
Additional plot config file to use a different state variable

bikegeek committed:
no longer needed

bikegeek committed:
no longer needed

bikegeek committed:
Fix the test that saves the plot in a different directory

bikegeek committed:
Remove text that requires the default config file

bikegeek committed:
use this yaml file for the User's Guide

bikegeek committed:
Updates include adding the config files and some additional text and formatting

bikegeek committed:
Update unit_tests.yaml

added installation of cartopy to support the physics tendency plots

bikegeek committed:
Update test_planview.py

skip these tests for CI, the data is too large

bikegeek committed:
Update requirements.txt

include cartopy for the physics tendency plots

bikegeek committed:
Update unit_tests.yaml

remove pip installation of cartopy.  Uncommented the cartopy entry in the requirements.txt and py310_requirements.txt instead

bikegeek committed:
Merge branch 'develop' into issue449_optimize_physics_tend_config
